### ISSUE TITLE
Warrior update

### ISSUE THREAD

**Issue by parkscs** [sneezymud/sneezymud#343]
We've got a few classes that are really lagging behind and I'd like to help with bringing warriors in-line with the other classes.  

**Existing Issues**
Warriors are supposed to be the masters of war and combat; however, at present, other classes (namely deikhans) outclass warriors at straight up combat, the one thing warriors should be the best at.  Rather than just saying "nerf deikhans", I think (and as we've discussed) the better option is to bring the other classes up to their level and then to look at buffing NPC's a bit across the board to compensate for the power gain in player classes.

Warriors have a lot of damage skills; however, these abilities are severely under-tuned and are never used (in part due to another issue I'll cover below).  Bash, trip, headbutt, bodyspam and so should have a place in the game (as at present this is really all warriors get other than passive abilities which improve their autoattacks.

Warriors are extremely constrained by movement points.  As an example, one of my characters can exhaust his movement points simply autoattacking, without being hungry and without using any abilities that drain movement points (kick, bash, etc.).  While other classes may be able to do the same, one difference is that other classes by and large have ways to deal with this issue.  To compound the issue for warriors, one of their few good abilities is berserk but it prevents, amongst other things, casting.  As such, even multi-class warriors who may be able to heal their own vitality cannot do so if they choose to use one of the class's few good abilities.

And as a final point, warriors are uninteresting to play.  A solo warrior has no ability to heal himself, travel around the game efficiently, buff himself, and in general has no utility abilities.  And as mentioned above, if you do try to use your completely undertuned kit, you simply end up draining your movement points more quickly.  With no way to regen their health or movement more quickly, playing a solo warrior is an agonizing experience and offers nothing when compared to other classes in the game.

**Proposed Changes**
While Deikhans bring party buffs with their auras (which, combined with chivalry, bring buffs that allow deikhans to hit more often ), Warriors would benefit from some abilities that allow them to secure hits on opponents more frequently.  My current thought is to give warriors an ability/abilities that allow them to ignore some amount (flat amount and/or percentage) of their enemy's armor when landing hits.  This buff(s) could come in a number of different ways, and one change we've discussed would be a combo system for chaining together abilities to achieve a certain effect.  While the particulars are still up in the air at this point, at the end of the day I think warriors should be the masters of armed combat.

With regard to the warrior class having lots of useless abilities, I think this should be approached in a number of different ways.  At a basic level though, abilities should have a reasonable chance of landing (not 100% but not 1%), they shouldn't drain a large amount of an incredibly limited resource (movement points at present), they should do something that's at least somewhat impactful and their drawbacks should be commensurate with their utility.   At present, some warrior abilities almost never land, while others have drawbacks that make them completely undesirable to use (e.g., deathstroke drains significant movement points and frequently gets you counterattacked for 100+ damage).  While some of these changes like a combo system are beyond my comfort level at the moment, I do feel comfortable updating their abilities that just need some needless restrictions removed, some numbers tweaked, or some utility added.  I do not think the answer is just to turn headbutt into atomize damage-wise or to turn bash into kinetic wave effectiveness-wise.  Ideally, each of the warrior's fighting moves would have some niche, whether that's a bit of added utility (a buff, a debuff, extra damage, etc.), as otherwise warriors will just spam the highest damage move in their kit (which leaves them very uninteresting to play).  A combo system could certainly help with this, but I think there's some other changes we make in the meantime that will still help to modernize the class.

To address their movement point issue, I propose adding a new ability to the brawling tree called conditioning.  This ability would simply be checked everywhere in the code that movement points are consumed (walking, autoattacking, using abilities, etc.) and a successful skill check could reduce (or sometimes negate) the movement drain.  This would give warriors a niche as no other class has such an ability and would directly address their movement consumption issue.

And to make them a bit more interesting, I think their kit could use a few new abilities.  At a minimum, we should give them 2h spec from the deikhan kit.  But I was also thinking about an advanced berserking ability that would modify the baseline berserk ability.  To me berserk is the best ability in the warrior's existing kit (even though it comes with significant limitations/risks) and I think it could be interesting to build a bit more of the kit around this ability for warriors willing to risk using it.  I think there's also the potential for a passive ability that lets them ignore a small amount of enemy armor, and then potentially we could tie a short duration, larger armor ignoring ability into the combo system.

And I'll add it as a separate item here even though I mentioned it above, but I think ultimately a combo system would make warriors very fun to play and give them a unique mechanic in the game.  For example, a combo could be something like landing a successful bash against an enemy and then a successful kick - and if you do that, you gain some effect.  This could be a small buff that lasts a few hours or a very strong buff that might just last a couple rounds.  It could also be a debuff to the opponent or a small bit of extra ability damage.  Really there are tons of possibilities with this type of mechanic and it could be limited by requiring the skills in the combo to succeed against the opponent.  This obviously requires us to first address the issues above relating to the warrior kit, but I think it could be a very cool mechanic that would set warriors apart and encourage full use of their kit. 

**Final Notes**
This is intended to start a conversation about the class and to get some of the things we've discussed down in a single post.  I'm sure I left out a few ideas we talked about and I'm sure we'll come up with a lot more in the future, but my goal here is really just to start the conversation.  In terms of coding, I'm happy to work through some of this and I would expect this work would be done in multiple small phases.  Right now I think implementing a combo system is beyond my comfort level with Sneezy, but I'd be happy to help with the design of the system and perhaps as my familiarity with the code base grows I can help out with the coding or take this on myself.  I'd like to start simple and small though, perhaps with reworking a couple of skills like bash and deathstroke to modernize these skills a bit and adding 2h spec to the soldiering kit so that they will share this ability with deikhans.  From there, assuming it receives positive feedback, I'd like to tackle the conditioning ability, perhaps along with a few more skill reworks.  And as a third wave, I think an advanced berserking mechanic could be pretty interesting, but exactly what that ability would give you I'm still brainstorming.  If anyone is still reading, thanks for making it to the end.  Looking forward to the conversation. :)

Best,
Boz/Casey

**Comment by agednotdead** [comment]
Don't forget about making late defense hones attainable.  I like what I see here. Bslam, grapple, kick, etc., those should all be usable and worthwhile endeavors. Moves are a limiting factor needlessly.  Pretty spot on with most of what your'e saying here.  Lbd should kick in and skills shouldn't start at just poor.  So much of the class is just a: not attainable, and b: won't/don't work when you actually attain them.

**Comment by Durhamite2118** [comment]
There is no doubt that, like deiks, warriors should have advance offense and advanced defense.

The combo idea is very cool, bash/trip/spin/bodyslam with kick/stomp that does 400% damage plus a chance to stun like glacial proc would be interesting.

**Comment by Durhamite2118** [comment]
.. headbutt or kneestrike having a chance to stun (like glacial but 1/2 the time )would be interesting

**Comment by wershlak** [comment]
So it comes down to this: Do we like Sneezy warriors? Do we like how they play and how they feel thematically? I never did so I've been wanting to change them drastically. It's not doable all at once and doesn't preclude making smaller changes first to get them in a more fun place now for bigger changes later. Still I think it's worth it to have some discussions about where we eventually want to get to before we start.

High level: Warriors (along with deiks) are the tanks of sneezy. However they are styled more like the D&D warriors of old in that they are designed to be kings of combat offensively and defensively. My vision is that offensively they should be below monks/thieves but should be slightly better than deikhans. defensively, I thought warriors would be best at avoiding/absorbing melee damage. Maybe deiks are better at absorbing magic damage due to protection being their thing. I feel like warriors should be the most versatile of all melee combatants capable of using all styles and weapon and armor types. Also, I feel like, as tanks, warriors should have the best control abilities and debuffs that keep the group safe. I feel like thieves and mages should have more offensive debuffs like stripping armor so the whole group does more damage whereas warriors would have debuffs that keep a mob from switching or make it harder to cast or reduce the ability of the mob to hit. 

So warriors == 3rd best melee damage, most control, best tanks.

This give interesting multi class combos. War/deik should offer some additional defense stuffs. Warrior/thief should offer more damage and utility. Warrior/monk should offer more offense and some interesting solo utility.

This ^ isn't totally true now but we can work towards it. :) 

- Because balance is a thing I feel like we need to think about how to give warriors opportunities to be either offensive or defensive situationally. With deikhans I added a guardians light a defensive ability that works on shield block and I added 2h specialization for more offense (but not when using a shield). Also, smite increases offensive capability but at the expense of a lot of defense. Not that I really succeeded mind you, but there was some thought that deiks would have to go 1h+shield or 2h to switch between offense and defense. For warriors berserk is an example of an ability you don't really want to use while tanking (at least in a  group). I want to rework how shield block works so we can add something to give warriors better chance to block than any other class. We maybe need to brainstorm some ideas here. Something like a reworked death stroke that does good damage but doesn't work if you have a shield. Yeah, yeah, right now tanks don't need shields. We need to make some changes to fix that soon. Don't let that trip you up. Design around the idea that basic game functionality actually works.

Think about how to make the game more engaging for warriors. 

So the idea I had was combos:
- opening strike: low level warrior ability. low cooldown and low move cost. adds a small debuff that increases warrior damage and warrior can exploit with finishers
- finishers: Consumes the opening strike debuf and are more powerful but also provide cool affects that are situalltional
  - death stroke: high damage finisher
  - guarded strike: lower damage debuff finisher, lowers mobs offense
  - renewing strike:  lower damage finisher that puts a buff on the warrior that regens some health

abilities for warriors can have a cost of movement, a command lockout, a health cost, a debuff on the user, a long reuse timer.

---

Done brainstorming

Here's the things I think we can add now in the shorter term:

- add 2h specialization
- add a hit all ability, whirlwind thing
- change death stroke to remove the return attack thing. Maybe make it only useable once per game hour or something
- remove restrictions on bash. If you are flying you can bash anyone woo!

**Comment by parkscs** [comment]
I think before we get to adding new skills that consume movement points, we need to address the issue of movement points.  Warriors have nothing that helps their regen or their base movement.  Monks get both - tons of base movement points plus yoginsa that quickly regens movement.  Clerics and deikhans have refresh or second wind.  Mages don't really use movement points.  Yet warriors and thieves tend to use movement points (warriors way more than thieves) yet they are the most handicapped by movement points to begin with.  

Again, as an example, one of my tri-class warriors can run out of movement points simply autoattacking, not hungry, not thirsty, not using any abilities.  Why would I ever consider using an ability that drains my movement points when I have no way to regen them more quickly and when simply autoattacking can drain them to 0 (in which case I'm in real trouble without help or potions).   Perhaps not in wave 1, but certainly in wave 2 of changes I think it'd be good to address this issue.

As for wave 1, I agree on 2h spec - that's an easy addition and one that makes total sense.  I'd like to rework deathstroke a bit to make it useful, remove some penalties and some of the cost, but still keep a bit of the risk/downside.  Whirlwind I think needs to happen (if only because we can add it to NPC warriors and beef them up >:-) but also it would be a fairly unique, albeit niche ability.  And finally, I agree on removing bash restrictions.  

I'd like to also tackle removing some rescue restrictions in wave 1 (which affects deikhans as well).  But currently, we can't rescue someone who is mounted... which makes no sense and is really clunky for deikhans.  This is especially true if your mount is flying, and you have to order it to land (and it won't listen half the time) and then dismount, just to be rescued.  This is a pointless restriction and should just be removed.  I'm also inclined to suggest opening up some other restrictions on rescue (Why can't we rescue npc's?  It might be suicidal, but why can't I rescue while berserking?) but that's perhaps another discussion.

**Comment by Durhamite2118** [comment]
Offensively, I'm not a huge fan of whirlwind for two reasons; 1) World of Warcraft mimic and 2) 99% of the fights are against 1 mob. What about a buff that increases the chance to crit hit or, like aura of vengeance, a chance to "double or triple strike".  2H spec is a must for warriors. 

Any version of a skill/buff that increases the movement pool is a must OR dropping the movement costs of all warrior skills by 1/2.  I like increasing the move pool.  Tank wise, defensive trance is pretty awesome but burns so many moves its not practical, I'd suggest we start with increase in movement pool and see how that effects play.

Finally, anything we can do to make the existing warrior skills that are never used better would be a good start, I like the combo concept but dont know how much of a lift that would be coding wise.  kneestrike, headbutt, stomp, bash, trip, spin, deathstroke are use 0.00% of the time.

Thanks for starting this thread and the discussion is great.

Finally, lets make all the warrior skills actually count.  It's great

**Comment by wershlak** [comment]
Sneezy will never be a WoW clone. However, I think we'd be crazy not to consider ideas for fun gameplay because it might be like WoW. The rage mechanic is the thing that defined WoW warriors for me though. Anything else we can find examples of elsewhere. I know muds in the 90s had a hit-all ability for warriors.

I think it's a good ability for sneezy warriors. Will you use it all that much? Probably not. It would be situational. Situationally useful abilities are better than we have now.

Besides, the biggest reason to add it is so mobs can use it against PCs :) 

--

Yeah we def need to add something to improve the movement point situation with warriors. Some ideas:
- a short duration buff that regens moves (even while fighting)
- an ability where the warrior pushes himself to exhaustion and sacrifices health for more moves.
- a passive ability to give them a larger pool of moves like monks have.
- a passive ability that will trigger a proc like effect the restores some moves on a hit.

**Comment by parkscs** [comment]
I agree at present it wouldn't be that useful for PCs but would be useful for NPCs.  And I think in the future it may well be useful for PCs - we've all been talking about designing new encounters, AOE could be a significant part of those new fights.  

As for movement, I agree adding a higher movement pool would help.  And it'd be easy to code - we just copy a monk ability.  But it homogenizes the classes in the game.  One option would be to have a conditioning ability that buffs their movement pool but maybe not as much as monks get, but also reduces movement spend on autoattacking/walking/abilities/etc. if they pass an ability check.  That would address the issue but also keep them unique in the game.

And I think separately, some warrior abilities just need their movement costs reduced.  I think all of that combined should address their issue of feeling movement starved, with the only thing missing being a way to regen.  On that...

I've been thinking about how to give warriors some kind of regen, and my current thought is to perhaps give them a massive buff to their hp/mv regen per tick.  This could be done with a new passive ability or it could be a fairly short duration buff they gain from landing a killing blow. - call it bloodlust or something along those lines perhaps  But regen (or lack thereof) has been an issue with the class since I can remember and it makes levelling a solo warrior incredibly cumbersome compared to other classes.

**Comment by parkscs** [comment]
Writing my reply here, rather than in the pull request to keep the discussion all in this thread.

I do not have a separate document for balance, and planning to really take it in phases and adjust as we go forward to deal with balance issues, especially because balance is a moving target as other classes see changes.  The first phases will be small and right now I'm currently testing a large rework of the deathstroke ability, which keeps in the risk (although I changed the algorithm) and makes it at least possible to land against end-game NPCs (again I changed the algorithm).  The damage is up a bit, the move cost is down, but all that comes in exchange for the skill having a lockout and it's no longer spammable.

As far as what warrior is going to look like at the end of the day, my goal is for it to look like a Sneezy warrior at the end of the day but with a functional kit.  The goal is to find a way to make it so that players want to use the existing warrior abilities and to give them a reasonable chance of success when doing so.  I really don't plan to add a ton to the kit itself and most of what I do plan to add I want to work with/augment the existing abilities.  But that kit inherently includes a fair bit of risk on the offense side, with berserk locking you in combat and making you aggro, deathstroke having a chance of counter attack, and I plan to keep just about all of that risk in the game but give a reasonable reward in exchange for taking the risk.  On the defensive side, I think they're more or less OK at the moment but could use a bit of a tune-up and some tweaking on certain abilities.  No concrete plans for defense just yet though but a few ideas.

As far as class balance and the end result, I would argue that the other classes right now are not balanced and to get there would require buffs and perhaps a rework of some other classes as well.  Mages vs. Deikhans at the moment are in totally different spots from a balance perspective, and we've started talking about some mage reworks as well for the future.  But as far as what I'm targeting with balance, it's for warriors to be on par with deikhans at the end of the day (and we can work to bring other classes up to that level).  My plan is to go slowly and tackle this in waves and continue to look at the balance from the earlier waves as I make updates.  I spend a fair bit of time testing this stuff locally but you can never really tell about balance until players get their hands on it and spend a lot of time using it in the field.

Once all classes are that that level NPCs will almost certainly need a tune up and we'll need new end-game zones to match that level of power, but that's a ways out.

**Comment by wershlak** [comment]
The only things on my list for warriors:
- Add 2h spec
- Rework/Fix the logic for shield block so that you get a consistent % to be modified by a stat/stats
- Then add a skill for warriors only that improves shield block percent
- When flying, allow an attacker to bash a non-flying victim
- Adjust trance so that it's dependent on AGI rather than DEX

--

Beyond these small fixes it's all BIG stuff. At least for me that's all I could think of since I'm not a big fan of warriors.

I know you're adding 2h spec now. Fix bash and trance if you want. I'd like to tackle shield block when I have a chance

It seems like you you have some medium range fixes in mind. Make death stroke useable, find something to help with moves, maybe add hitall. These seem good and totally appropriate. Go for it.

--

If we get to a big rework of warriors later we can just incorporate your stuff into it. You don't need to figure it all out now.

**Comment by drewhershey** [comment]
Gonna close this out in light of the big batch of warrior updates that went in over the last couple years.
